File: sorting_algorithms/selection_sort.py
def selection_sort(nums:list[int]) -> list[int]:
    for i in range(len(nums)):
        smallest_idx = i
        for j in range(i + 1, len(nums)):
            if nums[smallest_idx] > nums[j]:
                smallest_idx = j
        nums[smallest_idx],nums[i] = nums[i], nums[smallest_idx]
    return nums
        
sorted = selection_sort([2,1,4,6,4,3,2,8,9,2,4,5,6])
print(f"Sorted list: {sorted}")

File: selection_sort.go
package main

import "cmp"

func findSmallest[T cmp.Ordered](arr []T) int {
	smallestIdx := 0
	for i, val := range arr {
		if val < arr[smallestIdx] {
			smallestIdx = i
		}
	}
	return smallestIdx
}

func selectionSort[T cmp.Ordered](arr []T) []T {
	newArr := make([]T, 0, len(arr))
	copiedArr := make([]T, len(arr))
	copy(copiedArr, arr)

	for len(copiedArr) > 0 {
		idx := findSmallest(copiedArr)
		newArr = append(newArr, copiedArr[idx])
		copiedArr = append(copiedArr[:idx], copiedArr[idx+1:]...)
	}
	return newArr
}